plugins/core/hu.bme.mit.gamma.action.model/src/hu/bme/mit/gamma/action/util/ActionModelValidator.java

```diff
@@ -174,14 +185,85 @@ protected ActionModelValidator() {}
 	
 	public Collection<ValidationResultMessage> checkBranch(Branch branch) {
 		Collection<ValidationResultMessage> validationResultMessages = new ArrayList<ValidationResultMessage>();
-		// Block is empty
 		Expression guard = branch.getGuard();
-		if (!typeDeterminator.isBoolean(guard)) {
+		EObject container = branch.eContainer();
+		// check if container is a SwitchStatement
+		if (container instanceof SwitchStatement) {
+			SwitchStatement switchStatement = (SwitchStatement) container;
+			if (!typeDeterminator.equalsType(switchStatement.getControlExpression(), guard) && !(branch.getGuard() instanceof DefaultExpression)) {
+				validationResultMessages.add(new ValidationResultMessage(ValidationResult.ERROR,
+						"Type of control expression must be same type of guard in a SwitchStatement!",
```

Comment:
Do not put an exclamation mark at the end of the error message (put nothing there).